File: 01-introduction-and-complexity/02-amortized-analysis/python/amortized_demo.py
"""攤銷分析（Amortized Analysis）示範 / Amortized analysis demo.  # Module docstring (bilingual).

本單元用「動態陣列（dynamic array）擴容」示範攤銷 O(1)：  # Traditional Chinese overview with context.
- 平常 append 很便宜（只寫入 1 次）  # Explain the usual cheap case.
- 偶爾會擴容，需要搬移很多元素（看起來很貴）  # Explain the occasional expensive resize.
- 但把一長串操作的成本平均分攤後，平均仍是常數  # Explain the amortized conclusion.

We count deterministic "operation costs" (writes + copies) instead of wall-clock time.  # English summary of the counting model.
"""  # End of module docstring.

from __future__ import annotations  # Enable forward references in type hints (future-proofing).  

from dataclasses import dataclass  # Use dataclass to bundle per-operation results for tests/printing.  
import sys  # Read CLI arguments for custom append counts.  
from typing import List, Sequence  # Type hints for readability and tooling.  


def potential(size: int, capacity: int) -> int:  # Compute the potential Φ used by the potential method.  
    """位能函數 Φ / Potential function Φ.  # Short bilingual docstring.

    我們使用：Φ = 2·size - capacity + 1  # State the exact formula used in this unit.
    - 這個 +1 只是常數位移，讓初始狀態 (size=0, cap=1) 的 Φ = 0  # Explain why the +1 exists.
    - 位能法的攤銷成本：amortized = actual + (Φ_after - Φ_before)  # Define the amortized-cost formula.
    """  # End of docstring.
    if capacity < 1:  # Defensively reject invalid capacities to keep the formula meaningful.  
        raise ValueError("capacity must be >= 1")  # Signal invalid input to the caller.  
    return 2 * size - capacity + 1  # Return Φ(size, capacity) using integer arithmetic.  


@dataclass(frozen=True)  # Make step records immutable so tests cannot mutate them accidentally.  
class AppendStep:  # Record one append operation's costs and accounting/potential state.  
    index: int  # 1-based operation index in the sequence (for debugging and tables).  
    size_before: int  # Size before the append (number of stored elements).  
    capacity_before: int  # Capacity before the append (allocated slots).  
    copied: int  # Number of elements copied during resize (0 if no resize).  
    actual_cost: int  # Actual cost for this append: 1 write + copied moves.  
    phi_before: int  # Φ before the operation (potential method).  
    phi_after: int  # Φ after the operation (potential method).  
    amortized_cost: int  # Amortized cost via potential method: actual + ΔΦ.  
    bank_after: int  # Remaining credit after charging 3 and paying the actual cost.  


class AmortizedDynamicArray:  # A tiny dynamic array implementation instrumented for amortized analysis.  
    def __init__(self) -> None:  # Initialize an empty table with a small starting capacity.  
        self._size = 0  # Track the number of elements actually stored.  
        self._capacity = 1  # Start with capacity 1 so the first write is simple and deterministic.  
        self._data: List[int | None] = [None] * self._capacity  # Pre-allocate a fixed-size buffer.  

        self._bank = 0  # Track credits for the accounting method (start with 0 credit).  
        self._total_actual_cost = 0  # Track the total actual cost across all appends.  
        self._total_copies = 0  # Track total number of copied elements across resizes.  
        self._max_actual_cost = 0  # Track the maximum single-operation actual cost.  
        self._steps: List[AppendStep] = []  # Keep a per-append log for tests and optional printing.  

    @property
    def size(self) -> int:  # Expose size as a read-only property for callers/tests.  
        return self._size  # Return current size.  

    @property
    def capacity(self) -> int:  # Expose capacity as a read-only property for callers/tests.  
        return self._capacity  # Return current capacity.  

    @property
    def bank(self) -> int:  # Expose current bank balance (credits) for callers/tests.  
        return self._bank  # Return current bank balance.  

    @property
    def total_actual_cost(self) -> int:  # Expose total actual cost accumulated so far.  
        return self._total_actual_cost  # Return total actual cost.  

    @property
    def total_copies(self) -> int:  # Expose total number of copies performed during resizes.  
        return self._total_copies  # Return total copied elements.  

    @property
    def max_actual_cost(self) -> int:  # Expose the max single-operation actual cost observed.  
        return self._max_actual_cost  # Return the maximum actual cost.  

    @property
    def steps(self) -> Sequence[AppendStep]:  # Expose the per-step log as a read-only sequence.  
        return tuple(self._steps)  # Return an immutable snapshot to prevent accidental mutation.  

    def _resize(self, new_capacity: int) -> int:  # Grow the backing buffer and return the copy cost.  
        if new_capacity < self._size:  # Ensure the new capacity can hold existing elements.  
            raise ValueError("new_capacity must be >= size")  # Reject invalid resize requests.  
        if new_capacity < 1:  # Ensure capacity is never zero to keep invariants simple.  
            raise ValueError("new_capacity must be >= 1")  # Reject invalid capacities.  

        copied = 0  # Count how many element moves we perform during the resize.  
        new_data: List[int | None] = [None] * new_capacity  # Allocate a new fixed-size buffer.  
        for i in range(self._size):  # Copy exactly the current size elements into the new buffer.  
            new_data[i] = self._data[i]  # Copy one element (this models one "move" cost).  
            copied += 1  # Count one copy/move for amortized-cost measurement.  
        self._data = new_data  # Swap the buffer reference to complete the resize.  
        self._capacity = new_capacity  # Update capacity to match the new buffer size.  
        return copied  # Return the number of copied elements as the resize's copy cost.  

    def append(self, value: int) -> AppendStep:  # Append one value and return a detailed cost record.  
        phi_before = potential(self._size, self._capacity)  # Capture Φ before we mutate the structure.  
        copied = 0  # Default: no resize, so no copies are performed.  
        actual_cost = 1  # Always count 1 for writing the new element into the table.  

        if self._size == self._capacity:  # Resize exactly when the table is full (classic doubling rule).  
            new_capacity = self._capacity * 2  # Double the capacity to achieve amortized O(1) growth.  
            copied = self._resize(new_capacity)  # Perform the resize and count copies deterministically.  
            actual_cost += copied  # Add copy cost to the actual cost for this operation.  

        self._data[self._size] = value  # Store the new value into the next free slot.  
        self._size += 1  # Increase the logical size after insertion succeeds.  

        phi_after = potential(self._size, self._capacity)  # Capture Φ after the append completes.  
        amortized_cost = actual_cost + (phi_after - phi_before)  # Compute amortized cost using potential method.  

        self._bank += 3 - actual_cost  # Accounting method: charge 3, then pay the actual cost from credit.  
        if self._bank < 0:  # Credit should never go negative if the amortized charge is sufficient.  
            raise RuntimeError("bank went negative (accounting invariant violated)")  # Fail fast for debugging.  

        self._total_actual_cost += actual_cost  # Accumulate total actual cost across all operations.  
        self._total_copies += copied  # Accumulate total copies so we can validate aggregate bounds.  
        self._max_actual_cost = max(self._max_actual_cost, actual_cost)  # Update max observed cost.  

        step = AppendStep(  # Build an immutable record describing this append's costs and state.  
            index=len(self._steps) + 1,  # Use the next 1-based index for readability in logs.  
            size_before=self._size - 1,  # Size before insertion is the new size minus 1.  
            capacity_before=self._capacity if copied == 0 else self._capacity // 2,  # Recover pre-resize capacity.  
            copied=copied,  # Record how many elements were moved during this operation.  
            actual_cost=actual_cost,  # Record actual cost for this operation.  
            phi_before=phi_before,  # Record Φ before mutation.  
            phi_after=phi_after,  # Record Φ after mutation.  
            amortized_cost=amortized_cost,  # Record amortized cost via potential method.  
            bank_after=self._bank,  # Record bank balance after charging/paying for this operation.  
        )  # Close AppendStep construction.  
        self._steps.append(step)  # Append step to the log so tests can validate per-step invariants.  
        return step  # Return the step record for optional callers.  


@dataclass(frozen=True)  # Make summary immutable so printing/tests are deterministic.  
class SimulationSummary:  # A compact summary for a sequence of appends.  
    m: int  # Number of append operations performed.  
    final_size: int  # Final size after simulation (should equal m).  
    final_capacity: int  # Final capacity after simulation (power of two >= m, with this doubling rule).  
    total_actual_cost: int  # Total actual cost across all appends (writes + copies).  
    total_copies: int  # Total number of copied elements during all resizes.  
    max_actual_cost: int  # Maximum single-operation actual cost in the sequence.  
    final_bank: int  # Final credit balance when charging 3 per operation (accounting method).  


def simulate_appends(m: int) -> SimulationSummary:  # Run the demo for m appends and return a summary.  
    """模擬連續 append / Simulate repeated appends."""  # Short bilingual docstring.  
    if m < 0:  # Reject negative operation counts because they do not make sense.  
        raise ValueError("m must be >= 0")  # Signal invalid input to the caller.  

    table = AmortizedDynamicArray()  # Create a fresh dynamic array so results depend only on m.  
    for i in range(m):  # Perform m appends with deterministic values.  
        table.append(i)  # Append one value and record costs internally.  

    return SimulationSummary(  # Return a compact summary for printing/tests.  
        m=m,  # Record the requested operation count.  
        final_size=table.size,  # Expose the final size.  
        final_capacity=table.capacity,  # Expose the final capacity.  
        total_actual_cost=table.total_actual_cost,  # Expose total actual cost.  
        total_copies=table.total_copies,  # Expose total copies.  
        max_actual_cost=table.max_actual_cost,  # Expose max single-operation cost.  
        final_bank=table.bank,  # Expose final bank balance.  
    )  # Close SimulationSummary construction.  


def _is_power_of_two(x: int) -> bool:  # Check whether x is a power of two (used for validation).  
    return x > 0 and (x & (x - 1) == 0)  # Standard bit trick: powers of two have exactly one set bit.  


def format_summary_table(summaries: List[SimulationSummary]) -> str:  # Format multiple summaries into a printable table.  
    header = f"{'m':>6} | {'total':>8} | {'avg':>6} | {'max':>6} | {'cap':>6} | {'copies':>7} | {'bank':>6}"  # Table header line.  
    separator = "-" * len(header)  # Separator line to improve readability.  
    lines = [header, separator]  # Start with header + separator.  
    for s in summaries:  # Render each summary as one row.  
        avg = (s.total_actual_cost / s.m) if s.m > 0 else 0.0  # Compute average actual cost per op (safe for m=0).  
        lines.append(  # Append one aligned row.  
            f"{s.m:6d} | {s.total_actual_cost:8d} | {avg:6.2f} | {s.max_actual_cost:6d} | {s.final_capacity:6d} | {s.total_copies:7d} | {s.final_bank:6d}"  # Row string.  
        )  # Close row append.  
    return "\n".join(lines)  # Join lines into a printable multi-line string.  


def _parse_ms(argv: List[str]) -> List[int]:  # Parse CLI args into a list of m values.  
    if not argv:  # Use defaults when no args are provided.  
        return [1, 2, 4, 8, 16, 32]  # Default sizes mirror the previous unit for consistency.  
    return [int(x) for x in argv]  # Parse all args as integers (ValueError surfaces on invalid input).  


def main(argv: List[str]) -> int:  # CLI entry point that prints amortized-analysis summaries.  
    ms = _parse_ms(argv)  # Determine which operation counts to simulate.  
    summaries: List[SimulationSummary] = []  # Accumulate results so we can format them together.  
    for m in ms:  # Simulate each requested m independently.  
        summary = simulate_appends(m)  # Run the simulation and capture its summary.  
        if summary.final_size != m:  # Sanity check: final size should always equal number of appends.  
            raise RuntimeError("size mismatch (internal bug)")  # Fail fast if invariants are violated.  
        if not _is_power_of_two(summary.final_capacity):  # Validate the doubling rule created a power-of-two capacity.  
            raise RuntimeError("capacity is not a power of two (internal bug)")  # Fail fast for debugging.  
        if summary.final_capacity < max(1, m):  # Validate capacity is large enough to hold all appended elements.  
            raise RuntimeError("capacity too small (internal bug)")  # Fail fast for debugging.  
        summaries.append(summary)  # Store the validated summary for formatting.  
    print(format_summary_table(summaries))  # Print a formatted table so learners can inspect growth.  
    return 0  # Indicate success to the OS.  


if __name__ == "__main__":  # Execute only when run as a script, not when imported by tests.  
    raise SystemExit(main(sys.argv[1:]))  # Exit with the return code from main().  

File: 01-introduction-and-complexity/02-amortized-analysis/javascript/amortizedDemo.js
// 02 攤銷分析示範（JavaScript）/ Amortized analysis demo (JavaScript).  // Bilingual file header for this unit.
// We simulate a dynamic array that doubles capacity when full, and we count writes + copies.  // English overview of the model.

"use strict";  // Enable strict mode so accidental globals are avoided.  

function potential(size, capacity) {  // Compute Φ for the potential method.  
  if (capacity < 1) {  // Reject invalid capacities so invariants remain well-defined.  
    throw new RangeError("capacity must be >= 1");  // Fail fast with a clear message.  
  }  // Close validation block.  
  return 2 * size - capacity + 1;  // Use Φ = 2·size - capacity + 1 so Φ(0,1)=0.  
}  // End of potential.  

class AmortizedDynamicArray {  // Minimal dynamic array with instrumentation for amortized analysis.  
  constructor() {  // Initialize an empty table with capacity 1.  
    this._size = 0;  // Track logical size (stored elements).  
    this._capacity = 1;  // Track allocated capacity (slots).  
    this._data = new Array(this._capacity);  // Allocate a fixed-size backing store.  

    this._bank = 0;  // Accounting-method credit balance (charge 3 per append).  
    this._totalActualCost = 0;  // Aggregate total actual cost.  
    this._totalCopies = 0;  // Aggregate total copy count across resizes.  
    this._maxActualCost = 0;  // Track maximum single-operation actual cost.  
    this._steps = [];  // Keep a per-operation log for tests/inspection.  
  }  // Close constructor.  

  get size() {  // Expose size for callers/tests.  
    return this._size;  // Return current size.  
  }  // End of size getter.  

  get capacity() {  // Expose capacity for callers/tests.  
    return this._capacity;  // Return current capacity.  
  }  // End of capacity getter.  

  get bank() {  // Expose bank (credit) for callers/tests.  
    return this._bank;  // Return current bank balance.  
  }  // End of bank getter.  

  get totalActualCost() {  // Expose total actual cost for aggregate checks.  
    return this._totalActualCost;  // Return total actual cost.  
  }  // End of totalActualCost getter.  

  get totalCopies() {  // Expose total copies for aggregate checks.  
    return this._totalCopies;  // Return total copies.  
  }  // End of totalCopies getter.  

  get maxActualCost() {  // Expose maximum single-operation actual cost.  
    return this._maxActualCost;  // Return max actual cost.  
  }  // End of maxActualCost getter.  

  get steps() {  // Expose per-operation steps for tests.  
    return this._steps;  // Return step log array (tests treat it as read-only).  
  }  // End of steps getter.  

  _resize(newCapacity) {  // Resize the backing store and return the number of copies performed.  
    if (newCapacity < this._size) {  // Ensure we do not lose existing elements.  
      throw new RangeError("newCapacity must be >= size");  // Reject invalid resize requests.  
    }  // Close validation block.  
    if (newCapacity < 1) {  // Keep capacity positive.  
      throw new RangeError("newCapacity must be >= 1");  // Reject invalid capacities.  
    }  // Close validation block.  

    const newData = new Array(newCapacity);  // Allocate a new fixed-size array.  
    let copied = 0;  // Count how many elements we copy.  
    for (let i = 0; i < this._size; i += 1) {  // Copy exactly the stored elements.  
      newData[i] = this._data[i];  // Copy one element (one "move" in our cost model).  
      copied += 1;  // Increment copy counter.  
    }  // Close copy loop.  
    this._data = newData;  // Swap backing store.  
    this._capacity = newCapacity;  // Update capacity to match new store.  
    return copied;  // Return copy count as resize cost.  
  }  // End of _resize.  

  append(value) {  // Append one element and return a detailed step record.  
    const phiBefore = potential(this._size, this._capacity);  // Capture Φ before mutation.  
    const sizeBefore = this._size;  // Capture size before insertion.  
    const capacityBefore = this._capacity;  // Capture capacity before possible resize.  

    let copied = 0;  // Default: no resize means zero copies.  
    let actualCost = 1;  // Always pay 1 for writing the new element.  
    if (this._size === this._capacity) {  // Resize exactly when full (doubling rule).  
      copied = this._resize(this._capacity * 2);  // Resize and count copies.  
      actualCost += copied;  // Add copy cost to this operation's actual cost.  
    }  // Close resize branch.  

    this._data[this._size] = value;  // Write the new element to the next slot.  
    this._size += 1;  // Increase logical size after successful write.  

    const phiAfter = potential(this._size, this._capacity);  // Capture Φ after mutation.  
    const amortizedCost = actualCost + (phiAfter - phiBefore);  // Compute amortized cost via potential method.  

    this._bank += 3 - actualCost;  // Accounting method: charge 3 and pay actual cost from credit.  
    if (this._bank < 0) {  // Credit should never be negative with this amortized charge.  
      throw new Error("bank went negative (accounting invariant violated)");  // Fail fast for debugging.  
    }  // Close invariant check.  

    this._totalActualCost += actualCost;  // Accumulate total actual cost.  
    this._totalCopies += copied;  // Accumulate total copies.  
    this._maxActualCost = Math.max(this._maxActualCost, actualCost);  // Track maximum per-op cost.  

    const step = {  // Create a plain object step record (simple and serializable).  
      index: this._steps.length + 1,  // 1-based operation index.  
      sizeBefore: sizeBefore,  // Size before.  
      capacityBefore: capacityBefore,  // Capacity before.  
      copied: copied,  // Copies performed in this op.  
      actualCost: actualCost,  // Actual cost.  
      phiBefore: phiBefore,  // Φ before.  
      phiAfter: phiAfter,  // Φ after.  
      amortizedCost: amortizedCost,  // Amortized cost via potential method (should be 3).  
      bankAfter: this._bank,  // Bank after charging/paying.  
    };  // Close step object.  
    this._steps.push(step);  // Append step record to the log.  
    return step;  // Return step record to callers.  
  }  // End of append.  
}  // End of AmortizedDynamicArray.  

function simulateAppends(m) {  // Run m appends and return a summary object.  
  if (!Number.isInteger(m) || m < 0) {  // Validate input so simulation is well-defined.  
    throw new RangeError("m must be an integer >= 0");  // Signal invalid input.  
  }  // Close validation block.  

  const table = new AmortizedDynamicArray();  // Use a fresh table for deterministic results.  
  for (let i = 0; i < m; i += 1) {  // Perform m appends.  
    table.append(i);  // Append a deterministic value.  
  }  // Close loop.  

  return {  // Return a compact summary for printing/tests.  
    m: m,  // Number of operations.  
    finalSize: table.size,  // Final size.  
    finalCapacity: table.capacity,  // Final capacity.  
    totalActualCost: table.totalActualCost,  // Total actual cost.  
    totalCopies: table.totalCopies,  // Total copies.  
    maxActualCost: table.maxActualCost,  // Max per-op cost.  
    finalBank: table.bank,  // Final bank balance.  
  };  // Close summary object.  
}  // End of simulateAppends.  

function isPowerOfTwo(x) {  // Check whether x is a power of two.  
  return Number.isInteger(x) && x > 0 && (x & (x - 1)) === 0;  // Use bit logic for power-of-two detection.  
}  // End of isPowerOfTwo.  

function parseMs(argv) {  // Parse CLI args into a list of m values.  
  if (argv.length === 0) {  // Use defaults when no args are provided.  
    return [1, 2, 4, 8, 16, 32];  // Default sizes match other units for consistency.  
  }  // Close default branch.  
  return argv.map((s) => Number.parseInt(s, 10));  // Parse each token as an integer.  
}  // End of parseMs.  

function formatTable(summaries) {  // Format summaries into a fixed-width ASCII table.  
  const header = `${"m".padStart(6)} | ${"total".padStart(8)} | ${"avg".padStart(6)} | ${"max".padStart(6)} | ${"cap".padStart(6)} | ${"copies".padStart(7)} | ${"bank".padStart(6)}`;  // Header line.  
  const separator = "-".repeat(header.length);  // Separator line.  
  const lines = [header, separator];  // Start with header + separator.  
  for (const s of summaries) {  // Render one row per summary.  
    const avg = s.m > 0 ? s.totalActualCost / s.m : 0;  // Compute average safely for m=0.  
    lines.push(  // Append aligned row.  
      `${String(s.m).padStart(6)} | ${String(s.totalActualCost).padStart(8)} | ${avg.toFixed(2).padStart(6)} | ${String(s.maxActualCost).padStart(6)} | ${String(s.finalCapacity).padStart(6)} | ${String(s.totalCopies).padStart(7)} | ${String(s.finalBank).padStart(6)}`  // Row string.  
    );  // Close push call.  
  }  // Close row loop.  
  return lines.join("\n");  // Join lines with newlines for printing.  
}  // End of formatTable.  

function main(argv) {  // CLI entry point.  
  const ms = parseMs(argv);  // Determine which m values to simulate.  
  const summaries = [];  // Accumulate summary rows.  
  for (const m of ms) {  // Simulate each m.  
    const s = simulateAppends(m);  // Run simulation and collect summary.  
    if (s.finalSize !== m) {  // Sanity check size invariant.  
      throw new Error("size mismatch (internal bug)");  // Fail fast.  
    }  // Close check.  
    if (!isPowerOfTwo(s.finalCapacity)) {  // Validate power-of-two capacity invariant.  
      throw new Error("capacity not power-of-two (internal bug)");  // Fail fast.  
    }  // Close check.  
    if (s.finalCapacity < Math.max(1, m)) {  // Validate capacity bound.  
      throw new Error("capacity too small (internal bug)");  // Fail fast.  
    }  // Close check.  
    summaries.push(s);  // Store summary for printing.  
  }  // Close loop.  
  console.log(formatTable(summaries));  // Print the formatted table to stdout.  
  return 0;  // Return a success code for consistency with other languages.  
}  // End of main.  

module.exports = {  // Export functions/classes for tests and reuse.  
  potential: potential,  // Export potential function.  
  AmortizedDynamicArray: AmortizedDynamicArray,  // Export dynamic array class.  
  simulateAppends: simulateAppends,  // Export simulator function.  
  isPowerOfTwo: isPowerOfTwo,  // Export power-of-two helper.  
};  // End of exports.  

if (require.main === module) {  // Run the CLI only when executed directly.  
  process.exitCode = main(process.argv.slice(2));  // Use Node's exitCode so we can still flush console output.  
}  // Close main guard.  

